components/ai_chat/renderer/page_text_distilling.cc

```diff
@@ -0,0 +1,154 @@
+    ax::mojom::Role::kImage,
+    ax::mojom::Role::kLabelText,
+    ax::mojom::Role::kNavigation,
+    /* input elements */
+    ax::mojom::Role::kTextField,
+    ax::mojom::Role::kTextFieldWithComboBox,
+    ax::mojom::Role::kComboBoxSelect,
+    ax::mojom::Role::kListBox,
+    ax::mojom::Role::kListBoxOption,
+    ax::mojom::Role::kCheckBox,
+    ax::mojom::Role::kRadioButton,
+    ax::mojom::Role::kSlider,
+    ax::mojom::Role::kSpinButton,
+    ax::mojom::Role::kSearchBox,
+};
+
+void GetContentRootNodes(const ui::AXNode* root,
+                         std::vector<const ui::AXNode*>* content_root_nodes) {
+  std::queue<const ui::AXNode*> queue;
+  queue.push(root);
+  while (!queue.empty()) {
+    const ui::AXNode* node = queue.front();
+    queue.pop();
+    // If a main or article node is found, add it to the list of content root
+    // nodes and continue. Do not explore children for nested article nodes.
+    if (node->GetRole() == ax::mojom::Role::kMain ||
+        node->GetRole() == ax::mojom::Role::kArticle) {
+      content_root_nodes->push_back(node);
+      continue;
+    }
+    for (auto iter = node->UnignoredChildrenBegin();
+         iter != node->UnignoredChildrenEnd(); ++iter) {
+      queue.push(iter.get());
+    }
+  }
+}
+
+void AddContentNodesToVector(const ui::AXNode* node,
+                             std::vector<const ui::AXNode*>* content_nodes) {
+  if (base::Contains(kContentRoles, node->GetRole())) {
+    content_nodes->emplace_back(node);
+    return;
+  }
+  if (base::Contains(kRolesToSkip, node->GetRole())) {
+    return;
+  }
+  for (auto iter = node->UnignoredChildrenBegin();
+       iter != node->UnignoredChildrenEnd(); ++iter) {
+    AddContentNodesToVector(iter.get(), content_nodes);
+  }
+}
+
+void AddTextNodesToVector(const ui::AXNode* node,
+                          std::vector<std::u16string>* strings) {
+  const ui::AXNodeData& node_data = node->data();
+
+  if (base::Contains(kRolesToSkip, node_data.role)) {
+    return;
+  }
+
+  if (node_data.role == ax::mojom::Role::kStaticText) {
+    if (node_data.HasStringAttribute(ax::mojom::StringAttribute::kName)) {
+      strings->push_back(
+          node_data.GetString16Attribute(ax::mojom::StringAttribute::kName));
+    }
+    return;
+  }
+
+  for (const auto* child : node->children()) {
+    AddTextNodesToVector(child, strings);
+  }
+}
+
+}  // namespace
+
+void DistillPageText(
+    content::RenderFrame* render_frame,
+    base::OnceCallback<void(const absl::optional<std::string>&)> callback) {
+  LOG(ERROR) << "HELLO " << __FILE__;
+  LOG(ERROR) << __func__;
+  // ui::AXTreeID tree_id = render_frame->GetWebFrame()->GetAXTreeID();
+  auto snapshotter = render_frame->CreateAXTreeSnapshotter(
+      ui::AXMode::kWebContents | ui::AXMode::kHTML | ui::AXMode::kScreenReader);
+  ui::AXTreeUpdate snapshot;
+  snapshotter->Snapshot(
+      /* max_nodes= */ 9000,
+      /* timeout= */ {}, &snapshot);
+  ui::AXTree tree(snapshot);
```

Comment:
Can that happen in the middle of taking a snapshot or is the snapshot a sync operation? I don't see any callback in the snapshotter, which makes me assume it's synchronous. But perhaps it can change from another thread. Either way, do you have something in mind for checking `snapshot`?